cli/jobs/single-step/r/accidents/src/aml_utils.R

```diff
@@ -0,0 +1,92 @@
+
+library(mlflow)
+library(httr)
+library(later)
+library(tcltk2)
+
+new_mlflow_client.mlflow_azureml <- function(tracking_uri) {
+  host <- paste("https", tracking_uri$path, sep = "://")
+  get_host_creds <- function () {
+    mlflow:::new_mlflow_host_creds(
+      host = host,
+      token = Sys.getenv("MLFLOW_TRACKING_TOKEN"),
+      username = Sys.getenv("MLFLOW_TRACKING_USERNAME", NA),
+      password = Sys.getenv("MLFLOW_TRACKING_PASSWORD", NA),
+      insecure = Sys.getenv("MLFLOW_TRACKING_INSECURE", NA)
+    )
+  }
+  cli_env <- function() {
+    creds <- get_host_creds()
+    res <- list(
+      MLFLOW_TRACKING_USERNAME = creds$username,
+      MLFLOW_TRACKING_PASSWORD = creds$password,
+      MLFLOW_TRACKING_TOKEN = creds$token,
+      MLFLOW_TRACKING_INSECURE = creds$insecure
+    )
+    res[!is.na(res)]
+  }
+  mlflow:::new_mlflow_client_impl(get_host_creds, cli_env, class = "mlflow_azureml_client")
+}
+
+get_auth_header <- function() {
+    headers <- list()
+    auth_token <- Sys.getenv("MLFLOW_TRACKING_TOKEN")
+    auth_header <- paste("Bearer", auth_token, sep = " ")
+    headers$Authorization <- auth_header
+    headers
+}
+
+get_token <- function(host, exp_id, run_id) {
+    req_headers <- do.call(httr::add_headers, get_auth_header())
+    token_host <- gsub("mlflow/v1.0","history/v1.0", host)
+    token_host <- gsub("azureml://","https://", token_host)
+    api_url <- paste0(token_host, "/experimentids/", exp_id, "/runs/", run_id, "/token")
+    GET( api_url, timeout(getOption("mlflow.rest.timeout", 60)), req_headers)
+}
+
+
+fetch_token_from_aml <- function() {
+    print("Refreshing token")
+    tracking_uri <- Sys.getenv("MLFLOW_TRACKING_URI")
+    exp_id <- Sys.getenv("MLFLOW_EXPERIMENT_ID")
+    run_id <- Sys.getenv("MLFLOW_RUN_ID")
+    sleep_for <- 1
+    time_left <- 30
+    response <- get_token(tracking_uri, exp_id, run_id)
+    while (response$status_code == 429 && time_left > 0) {
+        time_left <- time_left - sleep_for
+        warning(paste("Request returned with status code 429 (Rate limit exceeded). Retrying after ",
+                    sleep_for, " seconds. Will continue to retry 429s for up to ", time_left,
+                    " second.", sep = ""))
+        Sys.sleep(sleep_for)
+        sleep_for <- min(time_left, sleep_for * 2)
+        response <- get_token(tracking_uri, exp_id)
+    }
+
+    if (response$status_code != 200){
+        error_response = paste("Error fetching token will try again after sometime: ", str(response), sep = " ")
+        print(error_response)
+    }
+
+    if (response$status_code == 200){
+        text <- content(response, "text", encoding = "UTF-8")
+        json_resp <-jsonlite::fromJSON(text, simplifyVector = FALSE)
+        json_resp$token
+        Sys.setenv(MLFLOW_TRACKING_TOKEN = json_resp$token)
+        print("Refreshing token done")
```

Comment:
i'd use message() instead of print() for your code's diagnostics

Reply:
Done